pkg/kubelet/kubelet_pods.go

```diff
@@ -1433,20 +1433,34 @@ func getPhase(spec *v1.PodSpec, info []v1.ContainerStatus) v1.PodPhase {
 
 // generateAPIPodStatus creates the final API pod status for a pod, given the
 // internal pod status.
-func (kl *Kubelet) generateAPIPodStatus(pod *v1.Pod, podStatus *kubecontainer.PodStatus) v1.PodStatus {
+func (kl *Kubelet) generateAPIPodStatus(pod *v1.Pod, podStatus *kubecontainer.PodStatus, isTerminated bool) v1.PodStatus {
 	klog.V(3).InfoS("Generating pod status", "pod", klog.KObj(pod))
 
 	// use the previous pod status, or the api status, as the basis for this pod
 	oldPodStatus, found := kl.statusManager.GetPodStatus(pod.UID)
 	if !found {
 		oldPodStatus = pod.Status
 	}
-	s := kl.convertStatusToAPIStatus(pod, podStatus, oldPodStatus)
+	s := kl.convertStatusToAPIStatus(pod, podStatus, oldPodStatus, isTerminated)
 
 	// calculate the next phase and preserve reason
 	allStatus := append(append([]v1.ContainerStatus{}, s.ContainerStatuses...), s.InitContainerStatuses...)
 	s.Phase = getPhase(&pod.Spec, allStatus)
 	klog.V(4).InfoS("Got phase for pod", "pod", klog.KObj(pod), "oldPhase", oldPodStatus.Phase, "phase", s.Phase)
+
+	// Perform a three-way merge between the statuses from the status manager,
+	// runtime, and generated status to ensure terminal status is correctly set.
+	if isTerminated && s.Phase != v1.PodFailed && s.Phase != v1.PodSucceeded {
```

Comment:
Thinking through a possible scenario:

1. Pod is running on both apiserver and in kubelet
2. Another client sets pod phase = Failed via apiserver
3. Config update sees it, passes it to pod worker, pod worker sets becomeTerminating = true
4. Pod worker loop enters syncTerminatingPod, calls this method with `false`
5. This loop is bypassed, s.phase is Running
6. s.phase Running is passed to pod sync handler (which is wrong)
7. pod.status.phase is Running, so we don't trigger illegal phase transition

So we're missing an condition on this check.  We have to explicitly check all three sources for pod terminal, and then preserve that.

The guarantee this check has to have is if any source (s.phase, pod.Status.Phase, or oldPodStatus.Phase) is terminal, s.Phase has to be terminal as well.  The current chunk does not guarantee that.

Reply:
s is calculated from `s := kl.convertStatusToAPIStatus(pod, podStatus, oldPodStatus, isTerminated)`. Is it possible for `s` to ever be terminal if either `pod` or `oldPodStatus` is non-terminal?

Reply:
convertStatusToAPIStatus ignores phase, and isTerminated would be false, so in the scenario I expected in 4 that s.phase is Running (but correct me if I'm wrong)

Reply:
> This loop is bypassed because s.phase is Running

This block is bypassed because `isTerminated` is false, do we just want to drop that requirement? I can test that.